### PR TITLE
[#176068662] Add lock on email validation process

### PR DESCRIPTION
**Pull request by AleDore**
This PR implements a lock mechanism for the email validation process, using Redis to store information about pending email validation processes. If a key-value pair is present into Redis for the provided fiscalCode, Controller returns immediately a `ResponseSuccessAccepted` without  overload io-functions-app, otherwise it normally starts emailValidationProcess ( and store data on Redis )

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -420,3 +423,14 @@ log.info(
   "JWT support token expiration set to %s seconds",
   JWT_SUPPORT_TOKEN_EXPIRATION
 );
+
+export const EMAIL_VALIDATION_PROCESS_TTL: NonNegativeInteger = NonNegativeInteger.decode(
+  process.env.EMAIL_VALIDATION_PROCESS_TTL
```
Comment (gunzip) [inline src/config.ts:428]: can you add the unit of measure in the name ie. EMAIL_VALIDATION_PROCESS_TTL_MS ?

File: src/controllers/__tests__/profileController.test.ts [hunk 2]
```diff
@@ -100,6 +100,8 @@ const badRequestErrorResponse = {
   type: undefined
 };
 
+const lockEmailValidationProcessTtl = 10 as NonNegativeInteger;
```
Comment (gunzip) [inline src/controllers/__tests__/profileController.test.ts:103]:
```suggestion
const lockEmailValidationProcessTtl = 10 as Seconds;
```

File: src/controllers/profileController.ts [hunk 3]
```diff
@@ -104,7 +115,57 @@ export default class ProfileController {
...
+          // check if an email validation process exists in Redis
+          this.sessionStorage.isEmailValidationProcessPending(user.fiscal_code),
+        toError
+      )
+        .chain(fromEither)
+        .foldTaskEither(
+          () => startEmailValidationProcessTask,
+          isEmailValidationProcessPending =>
+            isEmailValidationProcessPending
+              ? // email validation process already requested, return 202
```
Comment (gunzip) [inline src/controllers/profileController.ts:146]: I guess we have to change the api specs as well

File: src/controllers/profileController.ts [hunk 4]
```diff
@@ -4,21 +4,32 @@
...
+  ResponseErrorInternal,
+  ResponseSuccessAccepted
 } from "italia-ts-commons/lib/responses";
 import { ISessionStorage } from "src/services/ISessionStorage";
 
 import { InitializedProfile } from "../../generated/backend/InitializedProfile";
 import { Profile } from "../../generated/backend/Profile";
 import { ExtendedProfile as ExtendedProfileApi } from "../../generated/io-api/ExtendedProfile";
 
+import { EMAIL_VALIDATION_PROCESS_TTL } from "../config";
```
Comment (gunzip) [inline src/controllers/profileController.ts:32]: can we pass this value as a parameter to the controller?

File: src/config.ts [hunk 5]
```diff
@@ -420,3 +423,14 @@ log.info(
   "JWT support token expiration set to %s seconds",
   JWT_SUPPORT_TOKEN_EXPIRATION
 );
+
+export const EMAIL_VALIDATION_PROCESS_TTL: NonNegativeInteger = NonNegativeInteger.decode(
```
Comment (gunzip) [inline src/config.ts:427]: we should have a default value

File: src/controllers/profileController.ts (hunk 3 above)
Comment (AleDore) [reply src/controllers/profileController.ts]: What do you mean? `startEmailValidation` already returns 202 in case of success :)

Comment (gunzip) [reply src/controllers/profileController.ts]: ok!

**Review by gunzip — COMMENTED** [review]
lgtm, why draft?

**Comment by AleDore** [comment]
> lgtm, why draft?

I'm trying to test it locally :)

**Comment by AleDore** [comment]
> > lgtm, why draft?
> 
> I'm trying to test it locally :)

it works !

**Comment by balanza** [comment]
We finally decided to handle this mechanism in https://github.com/pagopa/io-functions-app/pull/132

**Pull request closed without merging**